src/librbd/api/Trash.cc

```diff
@@ -529,32 +529,48 @@ int Trash<I>::purge(IoCtx& io_ctx, time_t expire_ts,
 
   NoOpProgressContext remove_pctx;
   uint64_t list_size = to_be_removed.size(), i = 0;
-  for (const auto &entry_id : to_be_removed) {
-    r = librbd::api::Trash<I>::remove(io_ctx, entry_id, true, remove_pctx);
-    if (r < 0) {
-      if (r == -ENOTEMPTY) {
-        ldout(cct, 5) << "image has snapshots - these must be deleted "
-                      << "with 'rbd snap purge' before the image can be "
-                      << "removed." << dendl;
```

Comment:
This looks a little weird to me. Shouldn't we improve this to handle snap purge automatically on image remove?

Reply:
This is just a thought not related to this PR.

Reply:
I don't think so.  Since we don't do it automatically on `rbd rm` (where at least from user's POV trash isn't involved) why should we automatically do it on `rbd trash rm` or `rbd trash purge`?

Reply:
Or are you saying that we should consider moving away from requiring `rbd snap purge` in all cases?  That would need very careful consideration since this behaviour goes back pretty much to the dawn of rbd.

Reply:
Yes, I think that all troubles start with remove not allowed unless all snapshots are purged. Right now I don't see a good reason why this could not be allowed if snapshots are not protected.

But I am also not sure if we want to change this historic behavior. And I agree that `rbd trash rm` should behave the same as `rbd rm`. But `rbd trash purge` looks a little different. It is a higher level and is probably allowed to be more intelligent, and similarly how `rbd snap purge` removes all snapshots, `rbd trash purge` could be allowed to remove snapshots too.

Reply:
> Yes, I think that all troubles start with remove not allowed unless all snapshots are purged. Right now I don't see a good reason why this could not be allowed if snapshots are not protected.

It's not just snapshots though.   Remove is also not allowed if an image as part of a group, for example.

Reply:
I don't think `rbd trash purge` is a higher level.  I guess it depends on how you look at it, but in my view its job is to deal with image expiration times and pool usage thresholds.  Other than that, it should just "call" `rbd trash rm`.  In particular, if all images are expired, it should be equivalent to

```
for id in $(rbd trash ls --format=json | jq -r '.[].id'); do
    rbd trash rm $id; done
done
```

Today, the user has to resort to `rbd trash rm` because `rbd trash purge` doesn't try hard enough.  If `rbd trash purge` was made smarter we would have the reverse situation: `rbd trash purge` would suddenly become "stronger" than `rbd trash rm`.  Still confusing...

If we make a snapshot-related change, it should affect everything so that we are consistent across the entire librbd API.

Reply:
ack